Mathlib/RingTheory/ZMod/UnitsCyclic.lean

```diff
@@ -0,0 +1,713 @@
+theorem pow_one_add_mul_pow_prime_modeq
+    (p : ℕ) (hp : p.Prime) (hp2 : p ≠ 2) (c : ℤ) (m : ℕ) (hm : 1 ≤ m) :
+    Int.ModEq (p ^ (m + 2)) ((1 + c * p ^ m) ^ p) (1 + c * p ^ (m + 1)) := by
+  convert pow_add_mul_pow_modeq p hp hp2 1 c m hm
+  · simp
+  · simp
+
+-- Ireland & Rosen, 4.1, Corollary 1 of Lemma 3
+theorem one_add_mul_prime_pow_modeq
+    {p : ℕ} (hp : p.Prime) (hp2 : p ≠ 2) (a : ℤ) (n : ℕ) :
+    Int.ModEq (p ^ (n + 2)) ((1 + a * p) ^ (p ^ n)) (1 + a * p ^ (n + 1)) := by
+  induction n with
+  | zero => simp
+  | succ n hrec =>
+    nth_rewrite 2 [pow_succ]
+    rw [pow_mul]
+    have := pow_modEq_pow_succ_of_modEq_pow  p hp (n + 2) (Nat.le_add_left 1 (n + 1))
+      _ _ hrec
+    simp only [add_assoc, Nat.reduceAdd] at this ⊢
+    apply Int.ModEq.trans this
+    apply pow_one_add_mul_pow_prime_modeq p hp hp2
+    exact Nat.le_add_left 1 n
+
+/-- Ireland & Rosen, §4.1, Corollary 2 of Lemma 3 -/
+theorem orderOf_one_add_mul_prime {p : ℕ} (hp : p.Prime) (hp2 : p ≠ 2) (a : ℤ)
+    (ha : ¬ (p : ℤ) ∣ a) (n : ℕ) :
+    orderOf (1 + a * p : ZMod (p ^ (n + 1))) = p ^ n := by
+  letI : Fact (p.Prime) := ⟨hp⟩
+  by_cases hn : n = 0
+  · simp [hn]
+    convert mul_zero _
+    simp [natCast_zmod_eq_zero_iff_dvd, hn]
+  · obtain ⟨n, rfl⟩ := Nat.exists_eq_add_one_of_ne_zero hn
+    apply orderOf_eq_prime_pow
+    · rw [← Int.cast_natCast, ← Int.cast_mul, ← Int.cast_one, ← Int.cast_add,
+        ← Int.cast_pow, intCast_eq_intCast_iff]
+      intro H
+      apply ha
+      rw [← intCast_zmod_eq_zero_iff_dvd]
+      rw [← Int.cast_zero]
+      rw [intCast_eq_intCast_iff]
+      have := one_add_mul_prime_pow_modeq hp hp2 a n
+      simp [Nat.cast_pow, add_assoc] at H
+      replace this := Int.ModEq.trans (Int.ModEq.symm this) H
+      rw [← Nat.cast_pow, ← intCast_eq_intCast_iff, Int.cast_add, add_eq_left] at this
+      rw [← Int.cast_zero] at this
+      rw [intCast_eq_intCast_iff] at this
+      rw [← add_assoc n 1 1, pow_succ _ (n + 1), mul_comm a, Nat.cast_mul] at this
+      rw [← mul_zero ((p : ℤ) ^ (n + 1)), Nat.cast_pow] at this
+      apply Int.ModEq.mul_left_cancel' _ this
+      simp [hp.ne_zero]
+    · rw [← Int.cast_natCast, ← Int.cast_mul, ← Int.cast_one, ← Int.cast_add,
+        ← Int.cast_pow, intCast_eq_intCast_iff]
+      have := one_add_mul_prime_pow_modeq hp hp2 a (n + 1)
+      rw [add_assoc _ 1 2, add_comm 1 2, ← add_assoc, pow_succ] at this
+      replace this := Int.ModEq.of_mul_right _ this
+      rw [← Nat.cast_pow, ← intCast_eq_intCast_iff] at this
+      rw [add_assoc _ 1 1]
+      simp only [Nat.reduceAdd]
+      rw [←intCast_eq_intCast_iff]
+      rw [this]
+      simp only [Int.cast_add, add_eq_left, Int.cast_mul]
+      convert mul_zero _
+      rw [Int.cast_natCast]
+      exact natCast_self (p ^ (n + 2))
+
+/-- If p is an odd prime, then `(ZMod (p ^ n))ˣ` is cyclic for all n -/
+theorem isCyclic_units_of_prime_pow (p : ℕ) (hp : p.Prime) (hp2 : p ≠ 2) (n : ℕ) :
```

Comment:
I don't really feel like this warrants a 100-line proof. You have an element of order `p ^ n` from the previous lemma, and you have an element of order `p - 1` in Z/pZ. There ought to be a lemma involving exponents or orders under surjective group homomorphisms that gives you an element of order `p - 1` mod `p ^ (n + 1)`. And then you can combine them with `Commute.orderOf_mul_eq_mul_orderOf_of_coprime`.